File: SoftCode/FE/room-service-fe/src/app/models/room.ts
import { GenderPreference, PropertyType, RoomType } from "./enum";
import { RoomLocation } from "./room-location";

export interface Room {
    id ?: string;
    name ?: string;
    price ?: number;
    floor ?: number;
    roomSize ?: number;
    location ?: RoomLocation;
    hasFan ?:boolean;
    hasAirConditioner ?:boolean;
    hasParking ?:boolean;
     hasPrivateBathroom?: boolean;
    hasBalcony?: boolean;
    hasKitchen?: boolean;
    hasFridge?: boolean;
    hasWashingMachine?: boolean;
    hasTV?: boolean;
    hasWiFi?: boolean;
    hasElevator?: boolean;
    maxOccupants?: number;
    isPetFriendly?: boolean;
    isSmokingAllowed?: boolean;
    isSharedRoom?: boolean;
    genderPreference?: GenderPreference;

    roomType?: RoomType;
    propertyType?: PropertyType;

    distanceToCenter?: number;
    nearbyLandmarks?: string[];

    isUtilityIncluded?: boolean;
    depositRequired?: boolean;
    minStayMonths?: number;

    hasPhotos?: boolean;
    photoCount?: number;
    hasVideoTour?: boolean;

    verifiedListing?: boolean;

    availableFrom?: string;   // ISO string
    availableTo?: string;     // ISO string

    createdAt?: string;
    lastUpdated?: string;

    extraAttributes?: Record<string, any>; 


}

/*

"content": [
        {
            "id": "68a28bcf3db7c1f20d03afea",
            "name": "Cozy Studio Apartment",
            "price": 350.0,
            "floor": 3,
            "roomSize": 28.5,
            "location": {
                "country": null,
                "city": "Phnom Penh",
                "district": "Chamkarmon",
                "street": null,
                "fullAddress": null
            },
            "hasFan": true,
            "hasAirConditioner": true,
            "hasParking": false,
            "hasPrivateBathroom": true,
            "hasBalcony": true,
            "hasKitchen": true,
            "hasFridge": true,
            "hasWashingMachine": false,
            "hasTV": false,
            "hasWiFi": true,
            "hasElevator": true,
            "maxOccupants": 2,
            "isPetFriendly": false,
            "isSmokingAllowed": false,
            "isSharedRoom": false,
            "genderPreference": "NO_PREFERENCE",
            "roomType": "STUDIO",
            "propertyType": "APARTMENT",
            "distanceToCenter": 2.3,
            "nearbyLandmarks": [
                "university",
                "mall",
                "hospital"
            ],
            "isUtilityIncluded": true,
            "depositRequired": true,
            "minStayMonths": 3,
            "hasPhotos": true,
            "photoCount": 5,
            "hasVideoTour": false,
            "verifiedListing": true,
            "availableFrom": "2025-09-01T00:00:00",
            "availableTo": "2026-09-01T00:00:00",
            "createdAt": "2025-08-18T09:00:00",
            "lastUpdated": "2025-08-18T09:05:00",
            "extraAttributes": {
                "furnished": true,
                "internetSpeedMbps": 50,
                "security": "24/7"
            }
        }

        */

File: SoftCode/BE/room-service/src/main/java/com/piseth/java/school/roomservice/domain/Room.java
package com.piseth.java.school.roomservice.domain;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.geo.GeoJsonPoint;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexType;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexed;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Document("visitor_rooms")
@CompoundIndexes({
    @CompoundIndex(name = "idx_status_availableFrom", def = "{'status': 1, 'availableFrom': 1}"),
    @CompoundIndex(name = "idx_addr_province_district", def = "{'address.provinceCode': 1, 'address.districtCode': 1}")
})
public class Room {
	
  @Id
  private String id; // same as Room aggregate id

  private String ownerId;
  private String name;
  private String description;
  private Double price;
  private String currencyCode;
  private Integer floor;
  private Double roomSize;
  private String roomType;
  private String propertyType;

  private Address address;
  
  /**
   * GeoJSON Point for geo queries.
   * Mongo expects coordinates in order: [longitude, latitude].
   */

  @GeoSpatialIndexed(type = GeoSpatialIndexType.GEO_2DSPHERE)
  private GeoJsonPoint geoPoint;

  private Boolean hasFan;
  private Boolean hasAirConditioner;
  private Boolean hasParking;
  private Boolean hasPrivateBathroom;
  private Boolean hasBalcony;
  private Boolean hasKitchen;
  private Boolean hasFridge;
  private Boolean hasWashingMachine;
  private Boolean hasTV;
  private Boolean hasWiFi;
  private Boolean hasElevator;

  private Integer maxOccupants;
  private Boolean isPetFriendly;
  private Boolean isSmokingAllowed;
  private Boolean isSharedRoom;
  private String genderPreference;

  private Double distanceToCenter;
  private List<String> nearbyLandmarks;
  private Boolean isUtilityIncluded;
  private Boolean depositRequired;
  private Double depositAmount;
  private Integer minStayMonths;
  private String contactPhone;

  private List<String> photoObjectKeys;
  //private List<String> photoUrls;
  private String videoUrl;
  private Boolean verifiedListing;

  private String status;
  private LocalDateTime availableFrom;
  private LocalDateTime availableTo;

  private LocalDateTime createdAt;
  private LocalDateTime updatedAt;
  private String createdBy;
  private String updatedBy;

  private Map<String, Object> extraAttributes;

  @Indexed
  private LocalDateTime lastEventAt;
  private boolean deleted; // soft delete 

}